tealish/base.py

```diff
@@ -71,95 +72,67 @@ def _tealish(self) -> str:
     def tealish(self) -> str:
         return self._tealish()
 
-    def get_scope(self) -> Dict[str, Any]:
-        scope: Dict[str, Dict[str, Any]] = {
-            "slots": {},
-            "consts": {},
-            "blocks": {},
-            "functions": {},
-        }
+    def get_scope(self) -> Scope:
+        scope = Scope()
         for s in self.get_scopes():
-            scope["consts"].update(s["consts"])
-            scope["blocks"].update(s["blocks"])
-            scope["slots"].update(s["slots"])
-            scope["functions"].update(s["functions"])
+            scope.update(s)
         return scope
 
-    def get_scopes(self) -> List[Dict[str, Dict[str, Any]]]:
+    def get_scopes(self) -> List[Scope]:
         scopes = []
         s = self.get_current_scope()
         while True:
             scopes.append(s)
-            if s["parent"]:
-                s = s["parent"]
+            if s.parent is not None:
+                s = s.parent
             else:
                 break
         return scopes
 
-    def get_const(self, name: str) -> "Constant":
+    def get_const(self, name: str) -> Tuple[AVMType, Any]:
         consts = {}
         for s in self.get_scopes():
-            consts.update(s["consts"])
+            consts.update(s.consts)
         return consts[name]
 
     def get_slots(self) -> Dict[str, Any]:
         slots = {}
         for s in self.get_scopes():
-            slots.update(s["slots"])
+            slots.update(s.slots)
         return slots
 
-    def get_var(self, name: str) -> Tuple[Any, Any]:
+    def get_var(self, name: str) -> Optional[Tuple[int, VarType]]:
         slots = self.get_slots()
         if name in slots:
             return slots[name]
         else:
-            return (None, None)
-
-    def declare_var(self, name: str, type: Union[str, Tuple[str, str]]) -> int:
-        slot, _ = self.get_var(name)
-        if slot is not None:
-            raise Exception(f'Redefinition of variable "{name}"')
+            return None
 
+    def declare_var(self, name: str, type: Union[AVMType, Tuple[str, str]]) -> int:
         scope = self.get_current_scope()
+        scope.update(self.get_scope())
```

Comment:
This line fixed the issue with not catching the slots outside the `main` block but I'm not immediately sure _why_